File: lib/features/practice/domain/entities/learning_stats.dart
import '../../../vocabulary/domain/entities/cefr_level.dart';

final class LearningStats {
  const LearningStats({
    required this.dueCount,
    required this.masteredCount,
    required this.totalCount,
    required this.cefrBreakdown,
    required this.currentStreak,
    required this.weeklyLog,
  });

  final int dueCount;
  final int masteredCount;
  final int totalCount;
  final Map<CEFRLevel, int> cefrBreakdown;
  final int currentStreak;
  final Map<String, int> weeklyLog; // "yyyy-MM-dd" → words practiced
}

File: apps/web/src/lib/learningStats.ts
import type { VocabRecord } from "./vocabRecords";

type CefrLevel = VocabRecord["cefrLevel"];
const CEFR_LEVELS: CefrLevel[] = ["a1", "a2", "b1", "b2", "c1", "c2"];

export interface LearningStats {
  dueCount: number;
  masteredCount: number;
  totalCount: number;
  cefrBreakdown: Record<CefrLevel, number>;
}

// Ports StatsService.computeStats()'s mastered threshold exactly.
export const MASTERED_INTERVAL_THRESHOLD = 21;

function isDue(record: VocabRecord, now: Date): boolean {
  return record.nextReviewAt === null || new Date(record.nextReviewAt).getTime() <= now.getTime();
}

// Ports StatsService.computeStats()'s due/mastered/CEFR logic exactly —
// the streak/weeklyLog half lives in dailyActivity.ts instead, since that
// part is persisted (Firestore) rather than derived fresh every call.
export function computeLearningStats(records: VocabRecord[], now: Date = new Date()): LearningStats {
  let dueCount = 0;
  let masteredCount = 0;
  const cefrBreakdown = Object.fromEntries(CEFR_LEVELS.map((l) => [l, 0])) as Record<CefrLevel, number>;

  for (const r of records) {
    if (isDue(r, now)) dueCount++;
    if (r.sm2Interval >= MASTERED_INTERVAL_THRESHOLD) masteredCount++;
    cefrBreakdown[r.cefrLevel] = (cefrBreakdown[r.cefrLevel] ?? 0) + 1;
  }

  return { dueCount, masteredCount, totalCount: records.length, cefrBreakdown };
}
